rust/src/crypto.rs

```diff
@@ -0,0 +1,314 @@
+/// goes out of scope, its contents will be zeroed out.
+#[derive(AsRef, AsMut, From, Serialize, Deserialize, Eq, PartialEq, Clone)]
+pub struct KeySeed(box_::Seed);
+
+impl ByteObject for KeySeed {
+    fn from_slice(bytes: &[u8]) -> Option<Self> {
+        box_::Seed::from_slice(bytes).map(Self)
+    }
+
+    fn zeroed() -> Self {
+        Self(box_::Seed([0; box_::PUBLICKEYBYTES]))
+    }
+
+    fn as_slice(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+}
+
+/// A seed that can be used for signing key pair generation. When
+/// `KeySeed` goes out of scope, its contents will be zeroed out.
+#[derive(AsRef, AsMut, From, Serialize, Deserialize, Eq, PartialEq, Clone)]
+pub struct SignKeySeed(sign::Seed);
+
+/// Generate a new random signing key pair
+pub fn generate_sign_key_pair() -> (PublicSigningKey, SecretSigningKey) {
+    let (pk, sk) = sign::gen_keypair();
+    (PublicSigningKey(pk), SecretSigningKey(sk))
+}
+
+/// Deterministically derive a new signing key pair from the given
+/// seed
+pub fn derive_sign_key_pair_from_seed(seed: &SignKeySeed) -> (PublicSigningKey, SecretSigningKey) {
+    let (pk, sk) = sign::keypair_from_seed(seed.as_ref());
+    (PublicSigningKey(pk), SecretSigningKey(sk))
+}
+
+impl ByteObject for SignKeySeed {
+    fn from_slice(bytes: &[u8]) -> Option<Self> {
+        sign::Seed::from_slice(bytes).map(Self)
+    }
+
+    fn zeroed() -> Self {
+        Self(sign::Seed([0; sign::PUBLICKEYBYTES]))
+    }
+
+    fn as_slice(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+}
+
+/// Public key for asymmetric authenticated encryption
+#[derive(
+    AsRef,
+    AsMut,
+    From,
+    Serialize,
+    Deserialize,
+    Hash,
+    Eq,
+    Ord,
+    PartialEq,
+    Copy,
+    Clone,
+    PartialOrd,
+    Debug,
+)]
+pub struct PublicEncryptKey(box_::PublicKey);
+
+impl ByteObject for PublicEncryptKey {
+    fn zeroed() -> Self {
+        Self(box_::PublicKey([0_u8; box_::PUBLICKEYBYTES]))
+    }
+
+    fn as_slice(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+
+    fn from_slice(bytes: &[u8]) -> Option<Self> {
+        box_::PublicKey::from_slice(bytes).map(Self)
+    }
+}
+
+pub const SEALBYTES: usize = sealedbox::SEALBYTES;
+
+impl PublicEncryptKey {
+    /// Encrypt a message `m` with this public key. The resulting
+    /// ciphertext length is [`SEALBYTES`] + `m.len()`.
+    ///
+    /// The function creates a new key pair for each message, and
+    /// attaches the public key to the ciphertext. The secret key is
+    /// overwritten and is not accessible after this function returns.
+    pub fn encrypt(&self, m: &[u8]) -> Vec<u8> {
+        sealedbox::seal(m, self.as_ref())
+    }
+}
+
+impl SecretEncryptKey {
+    /// Decrypt the ciphertext `c` using this secret key and the
+    /// associated public key, and return the decrypted message.
+    ///
+    /// If decryption fails `Err(())` is returned.
+    pub fn decrypt(&self, c: &[u8], pk: &PublicEncryptKey) -> Result<Vec<u8>, ()> {
```

Comment:
when using the `public_key()` method then we don't have to pass it. but calling `public_key()` on a secret key recomputes the public key from the secret key which involves a multiplication on the elliptic curve. this can become quite costly if it is done very often. the alternative is to pass the precomputed public key by reference as an extra argument.